lib/hanami/events.rb

```diff
@@ -0,0 +1,180 @@
+require 'hanami/utils/class_attribute'
+begin
+  require 'concurrent'
+rescue LoadError
+  raise "Cannot find `concurrent-ruby' gem, please install it and retry."
+end
+
+module Hanami
+  # Event framework for Hanami
+  #
+  # @since x.x.x
+  module Events
+    # Backends for Hanami::Events
+    #
+    # @since x.x.x
+    module Backends
+      # Memory backend for Hanami::Events
+      #
+      # @since x.x.x
+      # @api private
+      class Memory
+        # Instantiate a new memory backend
+        #
+        # @return [Hanami::Events::Backends::Memory] a new instance
+        #
+        # @since x.x.x
+        # @api private
+        def initialize
+          @topics = Concurrent::Map.new { |h, k| h[k] = Concurrent::Array.new }
+        end
+
+        # Subscribe to a topic
+        #
+        # @param topic [String,Symbol] the topic
+        # @param subscriber [#call] the subscriber
+        #
+        # @since x.x.x
+        # @api private
+        def subscribe(topic, subscriber)
+          topics[topic.to_sym].push(subscriber)
+        end
+
+        # Broadcast an event for a given topic
+        #
+        # @param topic [String,Symbol] the topic
+        # @param event [Hanami::Events::Event] the event
+        #
+        # @since x.x.x
+        # @api private
+        def broadcast(topic, event)
+          topics[topic.to_sym].each do |subscriber|
+            subscriber.call(event)
+          end
+        end
+
+        private
+
+        # @since x.x.x
+        # @api private
+        attr_reader :topics
+      end
+    end
+
+    # Broadcasted Event
+    #
+    # @since x.x.x
+    class Event
+      # Instantiate a new event
+      #
+      # @param [#to_hash,#to_h] the payload
+      #
+      # @return [Hanami::Events::Event] a new instance
+      #
+      # @since x.x.x
+      # @api private
+      def initialize(payload)
+        @payload = Utils::Hash.new(payload).symbolize!.freeze
```

Comment:
Just whatever I want to broadcast. So instead of 

```ruby
event = Event.new(email: 'pascal@something.com')
email = event.payload[:email]
```

I could do

```ruby
event = Event.new('pascal@something.com')
email = event.payload
```